### DISCUSSION TITLE
MUI causing Hydration issues with simple styling. Remix:vite 2.7.2

### DISCUSSION THREAD

**Discussion by KeiranHines** [remix-run/remix#8913] [Q&A]
I have recently upgraded to remix:run@2.7.2 using vite. This has broken every single page when using MUI. It appears to be caused by any styling added  to the MUI components using the sx prop, possibly others as well. 
e.g. The below example as an index page will cause the following  errors
src:  
```
export default function Index() {
  return (
    <Box sx={{border: '1'}}>Test</Box>
  );
}
```
error 1: 
```
Warning: Expected server HTML to contain a matching <meta> in <head>.
meta
head
html
```
error 2: 
```
Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```

If I change the index to simply be a Box without the styling prop, the page renders error free.
```
export default function Index() {
  return (
    <Box>Test</Box>
  );
}
```
This issues only appeared after updating the remix:run 2.7.2 (with cloudflare pages) and node 20.11.1. It was not present on 2.6.x and node 18.17.x

Any help to diagnose and fix the issue would be appreciated.

**Comment by kiliman** [comment]
Was the previous working version also on Vite? Or did you just switch to Vite as of Remix v2.7.2?

**Reply by akomm** [reply]
Should this issue still persist with remix 2.9.2 and react 18.3.1?

What is interesting, the hydration errors only start to come when I add ?url to my css (tailwind) file. When I don't, the file is not loaded because the content is inlined as url. But for some unclear reason the styling of the page is correct and I get no hydration errors. Jut an error with the invalid css url (inlined content again). 

When I check network tab there is no other css loaded at any time, even after full reload. Inspecting the rendered content, but also the raw returned from server there is no style inlined or anything. How is the page styled? Firefox seems to be also confused, because it shows an empty `<html></html>` in inspector first. Only when I *select* individual element, the content appears in the inspector. As if the inspector lost the reference to the correct element in the DOM. I guess might be also related to the full document replacement...

The issue must be with the full document reload, because I get <title> mismatch error and adding css links (head too) causes issues to.

**Reply by akomm** [reply]
BTW. I'm currently not really in a hurry to upgrade to vite, but regularly make a check if it works.

**Reply by akomm** [reply]
regarding the canary recommendation. I've seen it below, but the one that solved the issue for some was version 18.3.0 canary and by now we have 18.3.1 stable. I'd prefer upgrading when I don't need to use canary. I hope its not related to the release policy change on the react side. I want to be able to *pin* a version of my libs so I can move them only when I need and not be dragged along by one lib and forced to adjust everything for it to work.

**Reply by kiliman** [reply]
Yes, previously React canary was going to be v18.3. However, there were some features that they wanted to push out ahead of RSC support, etc. So they published a minor v18.3 and canary was now the new v19. It's currently in Release Candidate (RC), so if you want to fix the hydration issues, you should install `react@rc` and `react-dom@rc`.

**Reply by akomm** [reply]
So with canary I don't have the error. But now fetcher.load makes full reloads when the loader returns async "json" not using the json() function. Very weird. I have reverted the thing for now and will check back in a month or two.

**Reply by MoSattler** [reply]
Still having the loading fallback issue with the latest canary, and a fresh install of 
```
npx create-remix@latest --template remix-run/remix/templates/spa
```
when deployed to GH pages. Anyone an idea how to solve this?

**Comment by KeiranHines** [comment]
The previous working version was not Vite. I switched to Vite as of Remix v2.7.2 coming from 2.6.2. 

Using Hydration fallback I managed to get a page to render, however, I feel like that is just a bandaid over a larger problem. I don't think I should need that on every page. 

Additionally I am also getting some wrangler error messages that are preventing some pages from working. They may be unrelated and an additional issue to resolve after the page can render successfully.

**Reply by markanthonyortega** [reply]
having the exact same issue and haven't been able to figure out a solution. I'm on 2.7.2 not using Vite but am also using simple styling in my mui components (sx)

**Reply by kommundsen** [reply]
I also have the same issue and would like to find a solution/workaround and get on with using vite.

**Reply by kiliman** [reply]
Have you tried installing React Canary? React 18.2 has issues with hydration on full documents.

See my template https://github.com/kiliman/remix-vite-template

**Reply by kommundsen** [reply]
Thank you, I'll try it out.

**Reply by kommundsen** [reply]
Thank you again @kiliman for the quick response. Your suggestion solved my issues.

**Reply by cliffordfajardo** [reply]
@kiliman -  whats the purpose of `noExternal: process.env.NODE_ENV === "production" ? [/^@mui\//] : [],` inside the vite.config.ts file? Why enable it for prod and not localhost development?

**Comment by mqs24d** [comment]
Having the same issue. Checkout the [Codesandbox](https://codesandbox.io/p/devbox/remix-vite-mui-9nwfp7?file=%2Fapp%2Froot.tsx%3A61%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt79s5640007356nmdxkxmqz%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt79s5640002356njdzl22ex%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt79s5640004356n8i66wfih%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt79s5640006356nfsos7d0s%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt79s5640002356njdzl22ex%2522%253A%257B%2522id%2522%253A%2522clt79s5640002356njdzl22ex%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clt79s5640006356nfsos7d0s%2522%253A%257B%2522id%2522%253A%2522clt79s5640006356nfsos7d0s%2522%252C%2522activeTabId%2522%253A%2522clt7a2ceu01c3356mhwn46h7y%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A5173%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clt7a2ceu01c3356mhwn46h7y%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clt79s5640004356n8i66wfih%2522%253A%257B%2522id%2522%253A%2522clt79s5640004356n8i66wfih%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt79s5640003356nfkq09hzo%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clt7a0xpp00tl356mxt5l42v5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clt7a0xqo000zdjge8zfffk71%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt79s5640003356nfkq09hzo%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

The entry.client.tsx / entry.server.tsx / root.tsx code is copied from the official [mui-remix-example](https://github.com/mui/material-ui/tree/master/examples/material-ui-remix-ts).

I compared the html generated in entry.server.tsx with the one displayed in the browser and it's identical (obviously ignoring the codesandbox script).

**Comment by JanHuege** [comment]
Same issue here. Many projects and companies rely on mui. This makes it impossible to use remix vite in these contexts.

**Reply by Alvazz** [reply]
+1

**Comment by kiliman** [comment]
I've created a sample repo showing Remix+MUI+Vite

⚡️ StackBlitz
https://stackblitz.com/github/kiliman/remix-vite-basic/tree/mui

https://github.com/kiliman/remix-vite-basic/tree/mui

**Reply by mqs24d** [reply]
Ok looks good so far... so the errors might be caused by the emotion cache stuff. Edit: i oversaw the suppressHydrationWarning on the body element. I adjusted my example (locally) and it works

**Reply by kiliman** [reply]
That `suppressHydrationWarning` is mainly for attributes being added to `<body>`. 

I'm using React Canary to prevent hydration issues with Remix _full document_ hydration.

**Reply by JanHuege** [reply]
It seems like defaults exports in material are a problem:

```
import { Typography } from "@mui/material"; // Works
import Typography from "@mui/material/Typography";// Breaks
```

Also importing static resources like logos is not working.
> solved: from docs (https://remix.run/docs/en/main/future/vite#fix-up-css-imports-referenced-in-links)

```
-import styles from "~/styles/dashboard.css";
+import styles from "~/styles/dashboard.css?url";

export const links = () => {
  return [
    { rel: "stylesheet", href: styles }
  ];
}
```

We were using meta imports with new URL instead of the vite way ?url

**Reply by KeiranHines** [reply]
I have been exclusively using the `import {import { Typography } from "@mui/material";` syntax with React 20.11.1 and I am consistently getting the Hydration issues.

**Comment by KeiranHines** [comment]
I have created an example repo [here](https://github.com/KeiranHines/remix-vite-cloudflare-mui)
I simply ran `npx create-remix@latest --template remix-run/remix/templates/vite-cloudflare` `npm install @mui/material @emotion/react @emotion/styled` and then added a `<Box>` element to the index page.

Tested on node 20.11.1

**Reply by mikebewell** [reply]
Just curious if any solution was ever discovered?

**Reply by yhan0704** [reply]
I am not sure if this answer is right for you.

Replaced react and react-dom on dependencies in package.jon.
    "react": "18.3.0-canary-01ab35a9a-20240228",
    "react-dom": "18.3.0-canary-01ab35a9a-20240228",

canary is one of the solution

**Reply by mqs24d** [reply]
canary was the solution... with react 18.2.0 it was not working. There are a few restrictions:

- default imports are not working so instead of `import Box from '@mui/material/Box'` you must `import { Box } from ...`
- we didn't got the advanced emotion cache setup working so we stick to the default https://emotion.sh/docs/ssr#default-approach

**Reply by mikebewell** [reply]
just tried it with 
```
    "react": "^19.0.0-beta-26f2496093-20240514",
    "react-dom": "^19.0.0-beta-26f2496093-20240514"
```
seems to be working fine

**Comment by javida1492** [comment]
I am also experiencing this hydration error, even with `"react": "18.3.0-canary-a870b2d54-20240314"` and` react-dom": "18.3.0-canary-a870b2d54-20240314"`. The code in particular essentially the example code for [FormControl component](https://mui.com/material-ui/react-select/#system-BasicSelect.tsx). If i rip out the `Select`, then it's fine so but im not sure how to work around this right now. Any help?

**Reply by SaraChicaD** [reply]
yes it's mostly the dropdowns!

**Comment by askneedle** [comment]
Have we found a solution for this? We started building with Remix, Vite, and MUI, and are now completely dependent on the app. The app has grown, and this hydration issue is killing our productivity.

**Reply by brad2014** [reply]
Fixes in underlying packages seem to be good, but they aren't yet used by Remix out of the box.
I am successfully running dev with the most recently released versions of MUI, React and Vite.  This involves overriding the dependency constraints of @remix-run/react 2.16.3.

```json
"dependencies": {
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.0",
    "@mui/icons-material": "^7.0.1", <-- released March 28, 2025
    "@mui/material": "^7.0.1",
    "@remix-run/node": "^2.16.3",
    "@remix-run/react": "^2.16.3",
    "@remix-run/serve": "^2.16.3",
    "react": "^19.1.0", <-- release March 28, 2025
    "react-dom": "^19.1.0"
  },
  "devDependencies": {
    "@remix-run/dev": "^2.16.3",
    "@types/react": "^19.0.12",
    "@types/react-dom": "^19.0.4",
    "typescript": "^5.8.2",
    "vite": "^6.2.3",
    "vite-tsconfig-paths": "^5.1.4"
  },
  "overrides": {
    "@remix-run/react": {
      "react": ">=18.2.0",  <-- instead of "^18.2.0"
      "react-dom": ">=18.2.0",
      "@types/react": ">=18.2.0"
    }
  },
```

**Comment by debojyotighosh** [comment]
This is still a problem as of today.
Did anyone find a solution to this?
